### PR TITLE
ddl: handle index key type in `CreateIndex`

### PR DESCRIPTION
**Pull request by leoppro**
### What problem does this PR solve? 

See: pingcap/parser#454

To support `CREATE [UNIQUE | FULLTEXT | SPATIAL] INDEX index_name` syntax, 
we remove the `Unique` field and add a new field named `KeyType` in `ast.CreateIndexStmt`

this pr changed the `ddl.DDL` interface and `ddl.CreateIndex` function to handle `KeyType`

We also rename `ast.AlterAlgorithm` to `ast.AlgorithmType `

### Check List 

Tests 

 - Unit test
Passed
 - Integration test
Passed
 - Manual test (add detailed scripts or steps below)
None

Code changes

 - Has exported function/method change
`ddl.CreateIndex`
 - Has exported variable/fields change
 - Has interface methods change
`ddl.DDL`
 - Has persistent data change

### CODE REVIEW

**Comment by leoppro** [comment]
@XuHuaiyu @zz-jason PTAL

**Comment by leoppro** [comment]
PTAL @winkyao

File: cmd/importer/parser.go [hunk 1]
```diff
@@ -275,13 +275,14 @@ func parseIndex(table *table, stmt *ast.CreateIndexStmt) error {
...
 	for _, indexCol := range stmt.IndexColNames {
 		name := indexCol.Column.Name.L
-		if stmt.Unique {
+		if stmt.KeyType == ast.IndexKeyTypeUnique {
 			table.uniqIndices[name] = table.findCol(table.columns, name)
-		} else {
+		} else if stmt.KeyType == ast.IndexKeyTypeNone {
 			table.indices[name] = table.findCol(table.columns, name)
+		} else{
+			return errors.Errorf("unsupported index type on column %s.%s", table.name, name)
```
Comment (winkyao) [inline cmd/importer/parser.go]: Will it break the compatibility?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -3209,8 +3209,14 @@ func getAnonymousIndex(t table.Table, colName model.CIStr) model.CIStr {
 	return indexName
 }
 
-func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, unique bool, indexName model.CIStr,
+func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast.IndexKeyType, indexName model.CIStr,
 	idxColNames []*ast.IndexColName, indexOption *ast.IndexOption, ifNotExists bool) error {
+
+	// not support Spatial and FullText index
+	if keyType == ast.IndexKeyTypeFullText || keyType == ast.IndexKeyTypeSpatial {
+		return errUnsupportedIndexType.GenWithStack("FULLTEXT and SPATIAL index is not supported")
```
Comment (winkyao) [inline ddl/ddl_api.go]: ditto

Comment (winkyao) [reply ddl/ddl_api.go]: IMHO, It's better to keep the behavior is consistent with before.

File: cmd/importer/parser.go (hunk 1 above)
Comment (leoppro) [reply cmd/importer/parser.go]: it won't. when TiDB executes SQLs like `CREATE FULLTEXT|SPATIAL INDEX XXXX`
The previous behavior was thrown an error from the parser. However the TiDB parser are supporting `FULLTEXT|SPATIAL` now, therefore we should throw an error when TiDB find an unsupported AST

**Pull request merged**